edx_notifications/stores/mongo/store_provider.py

```diff
@@ -0,0 +1,140 @@
+
+            if unread:
+                query_object.append({'user_notification.read_at': None})
+
+        if type_name:
+            pass
+            # query_object.append({'msg.msg_type': 'null'})
+            # collection = collection.filter(msg__msg_type=type_name)
+
+        if start_date:
+            query_object.append({'created': {"gte": start_date}})
+
+        if end_date:
+            query_object.append({'created': {"lte": end_date}})
+
+        collection = self.collection.find(*query_object)
+        return list(collection)
+
+    def get_notifications_for_user(self, user_id, filters=None, options=None):
+        """
+
+        :param user_id:
+        :param filters:
+        :param options:
+        :return:
+        """
+        self._get_prepaged_notifications(user_id, filters, options)
+
+    def mark_user_notifications_read(self, user_id, filters=None):
+        pass
+
+    def get_num_notifications_for_user(self, user_id, filters=None):
+        pass
+
+    def purge_expired_notifications(self, purge_read_messages_older_than, purge_unread_messages_older_than):
+        pass
+
+    def bulk_create_user_notification(self, user_msgs):
+        pass
+
+    def get_notification_for_user(self, user_id, msg_id):
+        pass
+
+    def save_user_notification(self, user_msg):
+        """
+        Create or Update the mapping of a user to a notification.
+        """
+
+        super(MongoNotificationStoreProvider, self).save_notification_message(user_msg.msg)
+
+        user_notification = self.collection.find({'user_id': user_msg.user_id})
+        if user_notification.count() == 0:
+            user_notification = self.collection.insert(
+                {
+                    'user_id': user_msg.user_id,
+                    'user_notification': [{
+                        'msg_id': user_msg.msg.id,
+                        'created': user_msg.created if user_msg.created else datetime.datetime.now(pytz.UTC),
+                        'modified': datetime.datetime.now(pytz.UTC),
+                        'user_context': user_msg.user_context if user_msg.user_context else None,
+                        'read_at': user_msg.read_at if user_msg.read_at else None
+                    }]
+                }
+            )
+        else:
+            user_notification = self.collection.update(
+                {'user_id': user_msg.user_id},
+                {'$push': {
+                    'user_notification':
+                        {
+                            'msg_id': user_msg.msg.id,
+                            'created': datetime.datetime.now(pytz.UTC),
+                            'modified': datetime.datetime.now(pytz.UTC),
+                            'user_context': user_msg.user_context if user_msg.user_context else None,
+                            'read_at': user_msg.read_at if user_msg.read_at else None
+                        }
+                }}
+            )
```

Comment:
I think we can use "upsert" semantics here so we don't need to do this existence test (which - since we are not locking databases - can lead to race conditions). See: http://api.mongodb.org/python/current/api/pymongo/collection.html